### PR TITLE
Add TLS 1.2 ECDSA ciphersuites to gRPC defaults

### PR DESCRIPTION
**Pull request by calderonth**
More and more PKI frameworks are generating certificates using
ECDSA keypairs, enabling the ciphersuites in gRPC avoids having
to define the environment variable.

### CODE REVIEW

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by jboeuf** [comment]
Thanks for the Pull Request. Could you please add a test with test ecdsa certs?

**Comment by jboeuf** [comment]
Thanks for the test. Unfortunately, it does not cover the gRPC implementation of TLS which is done in the `ssl_transport_security.c` file. Sorry I should have been more specific :(. 

For testing, you have 2 choices.
1. do another end to end test with ecdsa certs (like `test/core/end2end/fixtures/h2_ssl.c`).
2. wait for a PR by @yihuazhang that will import some TSI (transport security) level unit tests which will make things much easier to test. 

If you're not in a hurry for this feature, I would recommend option 2.

**Comment by jboeuf** [comment]
this is ok to test.

**Comment by calderonth** [comment]
Some of the failures are due to the fact the `boringssl` in third_party is not up to date with master and lacks an OpenSSL TLS compatibility sumbol as describe here:  https://boringssl.googlesource.com/boringssl/+/48e1d180a46b8caaf6b4619db07d078fb7cbcc60.

The symbol was dropped and then reintroduced, the version used by grpc master is in between, making some tests fail.

**Comment by jboeuf** [comment]
My recollection is that we are tracking the chromium branch of boringssl and it looks like the latest one has the fix: https://boringssl.googlesource.com/boringssl/+/chromium-stable/include/openssl/ssl.h#3851

@nicolasnoble, can we update?

**Comment by nicolasnoble** [comment]
Feel free to update the submodules in this pull request.

**Comment by jboeuf** [comment]
Sorry for the complicated flow but a few things here;
- We want to upgrade boring to the latest chromium-stable branch and master.
- Looks like the sanity tests are not happy. At least you should clang-format your code.
- The new boringssl does not seem to compile (maybe you'll have better luck with the chromium-stable branch): https://grpc-testing.appspot.com/job/gRPC_pull_requests_linux/6817/testReport/junit/(root)/aggregate_tests/run_tests_c_linux_dbg_native/

**Comment by calderonth** [comment]
Well this is more painful than expected ... bumping boringssl also included commits that changes the way they do tests. For instance: https://github.com/google/boringssl/commit/b228541129167901a682f302d2dc052546e380f8 switched the `asn1_test.cc` to be included in the `crypto_test` Gtest that isn't run by grpc anyway.
Can I just remove the `boringssl_asn1_test` target from the Makefile within grpc to make jenkins happy?

File: test/core/handshake/client_ssl.c [hunk 1]
```diff
@@ -147,20 +152,34 @@ static void server_thread(void *arg) {
...
+  if (SSL_CTX_use_PrivateKey_file(
+          ctx, ((load_ecdsa) ? SSL_EC_KEY_PATH : SSL_RSA_KEY_PATH),
+          SSL_FILETYPE_PEM) < 0) {
     ERR_print_errors_fp(stderr);
     abort();
   }
-  if (SSL_CTX_use_PrivateKey_file(ctx, SSL_KEY_PATH, SSL_FILETYPE_PEM) < 0) {
+
+  // Enable automatic DHE and ECDHE selection
+  if (!SSL_CTX_set_ecdh_auto(ctx, 1)) {
```
Comment (jboeuf) [inline test/core/handshake/client_ssl.c]: I would expect us to need a call to this function in ssl_transport_security.c as well. Were you able to run the end to end tests successfully on your machine?

**Comment by jboeuf** [comment]
On the security side, things look good to me, except maybe for a missing call in ssl_transport_security.c. 

@nicolasnoble can you please have a look at the tweaks in the build system?

File: test/core/handshake/client_ssl.c (hunk 1 above)
Comment (calderonth) [inline test/core/handshake/client_ssl.c]: It's not really needed as you are already calling `SSL_CTX_set_tmp_ecdh` in `populate_ssl_context`. I had to add the call to `SSL_CTX_set_ecdh_auto` for the test to enable it but the main TSI is all set.

Comment (jboeuf) [inline test/core/handshake/client_ssl.c]: sounds good. Thanks.

**Comment by calderonth** [comment]
Anything missing on my side for this?

**Comment by nicolasnoble** [comment]
If you are changing the boringssl submodule on purpose, you have to update the associated test to validate you haven't changed it by mistake. Sanity is not happy as a result: https://grpc-testing.appspot.com/job/gRPC_pull_requests_linux_opt/378/

**Comment by jtattermusch** [comment]
No action here for >6months. Are we going to pick up this change or can we close this?

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @calderonth isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Comment by calderonth** [comment]
I signed the new CLA

**Comment by jtattermusch** [comment]
There is a number of merge conflicts that need to be resolved first.

**Comment by jiangtaoli2016** [comment]
Any update on this PR?

**Comment by jiangtaoli2016** [comment]
@calderonth Do you still want to move on this PR. If so, could you please rebase and resolve the conflict. Thanks much for your contribution.

**Comment by calderonth** [comment]
I'll try to rebase the whole lot this week.

**Comment by calderonth** [comment]
This is a massive mess, lots of the files in conflict have changed dramatically, I'll give it another go after Easter.

**Comment by yashykt** [comment]
Is this PR going to be revived? We are going to mark this obsolete if there is no activity for another month. A new PR can be created at that time.

**Comment by jiangtaoli2016** [comment]
Since we already support ESDSA cipher suites now in gRPC. We can close this one.

**Pull request closed without merging**